examples/ecs/observers.rs

```diff
@@ -0,0 +1,198 @@
+
+    // Pre-register all our components, resource and event types.
+    world.init_resource::<SpatialIndex>();
+    world.init_component::<Mine>();
+    world.init_component::<TriggerMines>();
+    world.init_component::<Explode>();
+
+    // Observers are triggered when a certain event it fired, each event is represented by a component type.
+    // This observer runs whenever `TriggerMines` is fired, observers run systems which can be defined as closures.
+    world.observer(
+        |observer: Observer<TriggerMines>,
+         mines: Query<&Mine>,
+         index: Res<SpatialIndex>,
+         mut commands: Commands| {
+            // You can access the event data via the `Observer`
+            let trigger = observer.data();
+            // Access resources
+            for e in index.get_nearby(trigger.pos) {
+                // Run queries
+                let mine = mines.get(e).unwrap();
+                if mine.pos.distance(trigger.pos) < mine.size + trigger.radius {
+                    // And queue commands, including firing additional events
+                    // Here we fire the `Explode` event at entity `e`
+                    commands.event(Explode).entity(e).emit();
+                }
+            }
+        },
+    );
+
+    // Observers can also listen for events triggering for a specific component.
+    // This observer runs whenever the `Mine` component is added to an entity, and places it in a simple spatial index.
+    world.observer(
+        |observer: Observer<OnAdd, Mine>, query: Query<&Mine>, mut index: ResMut<SpatialIndex>| {
+            let mine = query.get(observer.source()).unwrap();
+            let tile = (
+                (mine.pos.x / CELL_SIZE).floor() as i32,
+                (mine.pos.y / CELL_SIZE).floor() as i32,
+            );
+            index.map.entry(tile).or_default().insert(observer.source());
+        },
+    );
+
+    // Since observers run systems you can also define them as standalone functions rather than closures.
+    // This observer runs whenever the `Mine` component is removed from an entity (including despawning it)
+    // and removes it from the spatial index.
+    world.observer(remove_mine);
+
+    // Now we spawn a set of random mines.
+    for _ in 0..1000 {
+        world
+            .spawn(Mine {
+                pos: Vec2::new(
+                    (rand::random::<f32>() - 0.5) * 1200.0,
+                    (rand::random::<f32>() - 0.5) * 600.0,
+                ),
+                size: 4.0 + rand::random::<f32>() * 16.0,
+            })
+            // Observers can also listen to events targeting a specific entity.
+            // This observer listens to `Explode` events targeted at our mine.
+            .observe(
+                |observer: Observer<Explode>, query: Query<&Mine>, mut commands: Commands| {
+                    // If an event is targeting a specific entity you can access it with `.source()`
+                    let source = observer.source();
```

Comment:
Targeted and untargeted events have their own specific uses, allowing for this ambiguity is bit weird to me :/
I thought this would take a form more like `|observer: Observer<Event>, source: In<ObserverSource>|` which would make un-/targeted events strongly typed, although at a cost.

I have no strong opinion about this matter so I'd love to hear someone else's. :)

Reply:
Yeah we don't have to support events without an entity source if we don't want to, no strong opinion from me either.

Reply:
I would like to make this strongly typed. To me, writing bug-free applications is far more important than micro-optimizing implementation details

Reply:
I don't think this has anything to do with optimizing implementation details. This is more related to how nuanced the API surface is. Adding an additional `In` parameter seems a bit strange (even ignoring the implications for the macro/traits where I'm currently using the `In` parameter for the observer) when everything else is configured by the observer's type signature.
What happens when you want an observer to listen to multiple events but only some of them are sourced? Do we detect that and panic? We now need somewhere to store which event types are sourced vs not-sourced and provide APIs to set that on dynamic event ids.
I can understand this being desirable but I want it to be known this adds non-trivial complexity.
I would rather enforce events have to have sources and consider adding events without sources as a follow-up.

Reply:
To be honest I'm not even convinced having events without targets has a compelling use case, I think I mostly just allowed it because it was easy to. You can always just target a placeholder entity of some kind.

Reply:
I remember entity id 0 was reserved for 'null' entity. If we can use this here and have untargeted events be just sugar over targeted at null entity, that'd be great!

Reply:
Yes, I believe I'm already using that in the case of an untargeted event.